### DISCUSSION TITLE
Failed to send message due to rate limiting

### DISCUSSION THREAD

**Discussion by smarthome-ha** [bbernhard/signal-cli-rest-api#527] [Q&A]
Hello all,

first of all, thanks for the good tutorials for the Signal Messenger AddOn.
I executed all steps from the Homeassiant.md an registered a new number successfully.

Unfortunetly I couldn't send a test message with this command:
`curl -X POST -H "Content-Type: application/json" -d '{"message": "Test", "number": "+49151XXXXXX", "recipients": ["+49170XXXXXX"]}' 'http://192.168.0.XX:8080/v2/send'`
The returned error message looks like that:
`{"error":"Failed to send message due to rate limiting. Use the attached challenge tokens to lift the rate limit restrictions via the '/v1/accounts/{number}/rate-limit-challenge' endpoint.","challenge_tokens":["XXX-XXX-..."]}`

Based on the command info of:
[https://bbernhard.github.io/signal-cli-rest-api/#/Accounts/post_v1_accounts__number__rate_limit_challenge](url)
I tried this command to level up the rate limit:
`curl -X POST -H "Content-Type: application/json" -d '{"captcha": ["signal-hcaptcha.XXXX"], "challenge_tokes": ["XXXX..."]}' 'http://192.168.0.XX:8080/v1/accounts/+49151XXXXXX/rate-limit-challenge'`
But with this answer:
`"error":"Couldn't process request - invalid request"}`

In general I surprised about the rate limiting, the log of the Signal Messenger  AddOn shows now high number of send requests:
`[GIN] 2024/04/24 - 20:00:21 | 201 |   1.630852284s |    192.168.0.XX | POST     "/v1/register/+49151XXXXXX"`
`[GIN] 2024/04/24 - 20:02:42 | 201 |   9.216935186s |    192.168.0.XX | POST     "/v1/register/+49151XXXXXX/verify/XXXXXX"`
`[GIN] 2024/04/24 - 20:05:41 | 201 |   6.223695112s |    192.168.0.XX | POST     "/v1/groups/+49151XXXXXX"`
`[GIN] 2024/04/24 - 20:30:25 | 200 |           79.441µs |    192.168.0.XX | GET      "/v1/about"`
`[GIN] 2024/04/24 - 20:30:25 | 429 |  581.942461ms |    192.168.0.XX | POST     "/v2/send"`
`[GIN] 2024/04/24 - 20:45:59 | 429 |  523.564851ms |    192.168.0.XX | POST     "/v2/send"`
`[GIN] 2024/04/25 - 18:17:47 | 429 |  855.536024ms |    192.168.0.XX | POST     "/v2/send"`
`[GIN] 2024/04/25 - 18:30:50 | 400 |  371.959978ms |    192.168.0.XX | POST     "/v1/accounts/+49151XXXXXX/rate-limit-challenge"`
`[GIN] 2024/04/25 - 18:33:54 | 400 |                  79µs |    192.168.0.XX | POST     "/v1/accounts/+49151XXXXXX/rate-limit-challenge"`
`[GIN] 2024/04/25 - 18:34:21 | 400 |           67.578µs |    192.168.0.XX | POST     "/v1/accounts/+49151XXXXXX/rate-limit-challenge"`
`[GIN] 2024/04/25 - 18:36:14 | 400 |           75.806µs |    192.168.0.XX | POST     "/v1/accounts/+49151XXXXXX/rate-limit-challenge"`

Do you have an idea in general, why I have this rate limit issue and how the command should look like, to extend this limit?
Thank you very much for your help in advance.

Used Version:
`Signal Messenger AddOn 0.83.0`
`Home Assistant Core 2024.4.3`

**Comment by bbernhard** [comment]
What mode are you using? 

Rate Limiting is done on the Signal Server side. Unfortunately I do not know what the actual conditions are that one runs into rate limiting issues.

The Rate Limiting can be lifted with the mentioned API endpoint - but you need to be careful, the rate limiting endpoints are itself rate limited.

**Reply by smarthome-ha** [reply]
Hello bbernhard,

thanks for the quick reply.

I use the Signal Messanger AddOn with the mode:
`json-rpc`

My guess is that the input to extend the limit is not correct. Do you have any experience as to whether the syntax is correct?
`curl -X POST -H "Content-Type: application/json" -d '{"captcha": ["signal-hcaptcha.XXXX"], "challenge_tokes": ["XXXX..."]}' 'http://192.168.0.XX:8080/v1/accounts/+49151XXXXXX/rate-limit-challenge'`
Because the error message suggests something like this.

**Reply by bbernhard** [reply]
Yes, the payload looks a bit wrong. It should be like this:

```
curl -X POST -H "Content-Type: application/json" -d '{"captcha": "<signal-captcha>", "challenge_token": "<challenge token>"}' 'http://192.168.0.XX:8080/v1/accounts/+49151XXXXXX/rate-limit-challenge'
```

If you open the [swagger documentation for that endpoint](https://bbernhard.github.io/signal-cli-rest-api/#/Accounts/post_v1_accounts__number__rate_limit_challenge) and click on the "Try it out" button you can enter your number and it will generate you the correct curl command.

**Comment by smarthome-ha** [comment]
Hello bbernhard,

the adapted curl-command for the rate limit extension works and after the extension the message send process also works fine.
Thank you very much for your quick and good support.
